AsyncTCP compile error for 2025.7.5 in LibreTiny for RTL

### The problem

Since last year, I had been using **libretiny version 1.6.0**, and up until version **2025.7.4,** the device compiled without any errors. Starting with version 2025.7.5, I began encountering compilation errors. I thought that perhaps I needed to switch to the latest version of **libretiny.** I updated the configuration, but that didn’t help—I still couldn’t compile.

### Which version of ESPHome has the issue?

2025.7.5

### What type of installation are you using?

Home Assistant Add-on

### What platform are you using?

RTL87XX

### Component causing the issue

_No response_

### YAML Config

```yaml
substitutions:
  devicename: smartplug
  friendly_name: t102-v1.1-2
  device_description: Digitaling 20A Power Monitor RTL8710BX t102-v1.1
  current_res: "0.001" # Power monitoring calibration https://esphome.io/components/sensor/hlw8012.html
  voltage_div: "1600" # Power monitoring calibration


esphome:
  name: ${devicename}
  friendly_name: ${friendly_name}
  comment: ${device_description}

rtl87xx:
  board: wr2 # actually t102-v1.1 but https://github.com/libretiny-eu/libretiny/issues/247
  framework:
    version: latest

# Enable logging
logger:
  level: INFO

# Enable Home Assistant API
api:
  encryption:
    key: "nkFfyNiXCADTQtOQa8Zjc76LcomhpTG237vDXBuHusM="

ota:
  - platform: esphome
    password: "a12a4ea51c42d07be2abccab2a3ed680"

wifi:
  ssid: !secret wifi_ssid2
  password: !secret wifi_password
  fast_connect: true

  # Enable fallback hotspot in case wifi connection fails
  ap:
    ssid: ${friendly_name} Fallback Hotspot
    password: "7zKaNjgJ2O7Q"

captive_portal:

web_server:
  version: 1 # version 2/3 doesn't work

# Enable time component for use by daily power sensor
time:
  - platform: homeassistant
    id: homeassistant_time

binary_sensor:
# Button on the front is pressed and then toggle relay
  - platform: gpio
    device_class: power
    pin:
      number: PA18
      mode: INPUT_PULLUP
      inverted: True
    name: Button
    on_press:
      - switch.toggle: relay
      - switch.toggle: statusled

text_sensor:
# Reports the ESPHome Version with compile date
  - platform: version
    name: ESPHome Version
  - platform: libretiny
    version:
      name: LibreTiny Version

switch:
# Relay itself
  - platform: gpio
    name: Relay
    pin: PA15
    id: relay
    restore_mode: RESTORE_DEFAULT_OFF
  - platform: gpio
    name: statusled
    pin: PA05
    id: statusled
    restore_mode: RESTORE_DEFAULT_OFF
    inverted: true

#bootloop fix: https://github.com/libretiny-eu/libretiny/issues/155#issuecomment-1826470433
#https://esphome.io/components/sensor/hlw8012.html
sensor:
  - platform: hlw8012
    model: BL0937
    sel_pin:
       number: PA14
       inverted: true
    cf_pin: PA00
    cf1_pin: PA12
    current:
      name: "Current"
    voltage:
      name: "Voltage"
    power:
      name: "Power"
      id: power
    update_interval: 1s
    initial_mode: CURRENT
    change_mode_every: 8
    current_resistor: ${current_res}
    voltage_divider: ${voltage_div}
# Reports the total Power so-far each day, resets at midnight, see https://esphome.io/components/sensor/total_daily_energy.html
  - platform: total_daily_energy
    name: Total Daily Energy
    icon: mdi:circle-slice-3
    power_id: power
    filters:
      - multiply: 0.001
    unit_of_measurement: kWh
    state_class: total_increasing
    device_class: energy
```

### Anything in the logs that might be useful for us?

```txt
INFO ESPHome 2025.7.5
INFO Reading configuration /config/esphome/rtl8710bx.yaml...
INFO Detected timezone 'Europe/Berlin'
INFO Generating C++ source...
INFO Compiling app...
Processing smartplug (board: wr2; framework: arduino; platform: libretiny)
--------------------------------------------------------------------------------
Library Manager: Installing ESP32Async/AsyncTCP @ 3.4.5
INFO Installing ESP32Async/AsyncTCP @ 3.4.5
Unpacking  [####################################]  100%
Library Manager: AsyncTCP@3.4.5 has been installed!
INFO AsyncTCP@3.4.5 has been installed!
Library Manager: Installing ESP32Async/ESPAsyncWebServer @ 3.7.10
INFO Installing ESP32Async/ESPAsyncWebServer @ 3.7.10
Unpacking  [####################################]  100%
Library Manager: ESPAsyncWebServer@3.7.10 has been installed!
INFO ESPAsyncWebServer@3.7.10 has been installed!
Library Manager: Resolving dependencies...
INFO Resolving dependencies...
Library Manager: Installing esphome/noise-c @ 0.1.10
INFO Installing esphome/noise-c @ 0.1.10
Unpacking  [####################################]  100%
Library Manager: noise-c@0.1.10 has been installed!
INFO noise-c@0.1.10 has been installed!
Library Manager: Resolving dependencies...
INFO Resolving dependencies...
Library Manager: Installing esphome/libsodium @ 1.10020.7
INFO Installing esphome/libsodium @ 1.10020.7
Unpacking  [####################################]  100%
Library Manager: libsodium@1.10020.7 has been installed!
INFO libsodium@1.10020.7 has been installed!
Library Manager: Installing bblanchon/ArduinoJson @ 7.4.2
INFO Installing bblanchon/ArduinoJson @ 7.4.2
Unpacking  [####################################]  100%
Library Manager: ArduinoJson@7.4.2 has been installed!
INFO ArduinoJson@7.4.2 has been installed!
Warning! Non-Git installations are NOT SUPPORTED.
HARDWARE: RTL8710BN 125MHz, 256KB RAM, 788KB Flash
 - framework-arduino-api @ 2022.8.24+sha.237b10a 
 - framework-realtek-amb1 @ 0.0.0+v2022.06.21.sha.c4e44ef 
 - library-flashdb @ 1.2.0+sha.d5c892f 
 - library-freertos @ 8.1.2+sha.776ae6c 
 - library-freertos-port @ 2025.1.18+sha.dbfc408 
 - library-lwip @ 2.1.3-amb1+sha.6297b80 
 - library-printf @ 6.1.0+sha.28a79bd 
 - library-uf2ota @ 5.0.0+sha.f955412 
PLATFORM VERSIONS:
 - libretiny @ 1.9.1
 - ltchiptool @ 4.12.1
CUSTOM OPTIONS:
 - fw_name = esphome
 - fw_version = 2025.7.5
Library Manager: Installing DNSServer
INFO:Library Manager:Installing DNSServer
Unpacking 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100%
Library Manager: DNSServer@1.1.0 has been installed!
INFO:Library Manager:DNSServer@1.1.0 has been installed!
Dependency Graph
|-- AsyncTCP @ 3.4.5
|-- ESPAsyncWebServer @ 3.7.10
|-- DNSServer @ 1.1.0
|-- noise-c @ 0.1.10
|-- ArduinoJson @ 7.4.2
Compiling .pioenvs/smartplug/src/esphome/components/api/api_connection.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/api/api_frame_helper.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/api/api_pb2.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/api/api_pb2_service.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/api/api_server.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/api/list_entities.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/api/proto.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/api/subscribe_state.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/binary_sensor/automation.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/binary_sensor/binary_sensor.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/binary_sensor/filter.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/captive_portal/captive_portal.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/esphome/ota/ota_esphome.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/gpio/binary_sensor/gpio_binary_sensor.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/gpio/switch/gpio_switch.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/hlw8012/hlw8012.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/homeassistant/time/homeassistant_time.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/json/json_util.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/libretiny/core.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/libretiny/gpio_arduino.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/libretiny/helpers.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/libretiny/lt_component.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/libretiny/preferences.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/logger/logger.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/logger/logger_libretiny.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/md5/md5.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/mdns/mdns_component.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/mdns/mdns_libretiny.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/network/util.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/ota/ota_backend.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/ota/ota_backend_arduino_libretiny.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/pulse_counter/pulse_counter_sensor.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/safe_mode/safe_mode.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/sensor/automation.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/sensor/filter.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/sensor/sensor.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/socket/lwip_sockets_impl.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/socket/socket.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/switch/automation.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/switch/switch.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/text_sensor/filter.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/text_sensor/text_sensor.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/time/automation.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/time/real_time_clock.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/total_daily_energy/total_daily_energy.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/version/version_text_sensor.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/web_server/list_entities.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/web_server/ota/ota_web_server.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/web_server/web_server.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/web_server/web_server_v1.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/web_server_base/web_server_base.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/wifi/wifi_component.cpp.o
Compiling .pioenvs/smartplug/src/esphome/components/wifi/wifi_component_libretiny.cpp.o
Compiling .pioenvs/smartplug/src/esphome/core/application.cpp.o
Compiling .pioenvs/smartplug/src/esphome/core/color.cpp.o
Compiling .pioenvs/smartplug/src/esphome/core/component.cpp.o
Compiling .pioenvs/smartplug/src/esphome/core/component_iterator.cpp.o
Compiling .pioenvs/smartplug/src/esphome/core/controller.cpp.o
Compiling .pioenvs/smartplug/src/esphome/core/entity_base.cpp.o
Compiling .pioenvs/smartplug/src/esphome/core/helpers.cpp.o
Compiling .pioenvs/smartplug/src/esphome/core/log.cpp.o
Compiling .pioenvs/smartplug/src/esphome/core/scheduler.cpp.o
Compiling .pioenvs/smartplug/src/esphome/core/string_ref.cpp.o
Compiling .pioenvs/smartplug/src/esphome/core/time.cpp.o
Compiling .pioenvs/smartplug/src/esphome/core/util.cpp.o
Compiling .pioenvs/smartplug/src/main.cpp.o
Compiling .pioenvs/smartplug/lib052/AsyncTCP/AsyncTCP.cpp.o
Compiling .pioenvs/smartplug/liba5f/ESPAsyncWebServer/AsyncEventSource.cpp.o
Compiling .pioenvs/smartplug/liba5f/ESPAsyncWebServer/AsyncJson.cpp.o
Compiling .pioenvs/smartplug/liba5f/ESPAsyncWebServer/AsyncMessagePack.cpp.o
Compiling .pioenvs/smartplug/liba5f/ESPAsyncWebServer/AsyncWebHeader.cpp.o
.piolibdeps/smartplug/AsyncTCP/src/AsyncTCP.cpp: In function 'void _send_async_event(lwip_tcp_event_packet_t*)':
.piolibdeps/smartplug/AsyncTCP/src/AsyncTCP.cpp:202:3: error: 'xTaskNotifyGive' was not declared in this scope
  202 |   xTaskNotifyGive(_async_service_task_handle);
      |   ^~~~~~~~~~~~~~~
.piolibdeps/smartplug/AsyncTCP/src/AsyncTCP.cpp: In function 'void _prepend_async_event(lwip_tcp_event_packet_t*)':
.piolibdeps/smartplug/AsyncTCP/src/AsyncTCP.cpp:210:3: error: 'xTaskNotifyGive' was not declared in this scope
  210 |   xTaskNotifyGive(_async_service_task_handle);
      |   ^~~~~~~~~~~~~~~
.piolibdeps/smartplug/AsyncTCP/src/AsyncTCP.cpp: In function 'void _async_service_task(void*)':
.piolibdeps/smartplug/AsyncTCP/src/AsyncTCP.cpp:322:5: error: 'ulTaskNotifyTake' was not declared in this scope
  322 |     ulTaskNotifyTake(pdTRUE, ASYNC_TCP_MAX_TASK_SLEEP);
      |     ^~~~~~~~~~~~~~~~
*** [.pioenvs/smartplug/lib052/AsyncTCP/AsyncTCP.cpp.o] Error 1
```

### Additional information

_No response_

Compiling .pioenvs/smartplug/liba5f/ESPAsyncWebServer/AsyncWebHeader.cpp.o
.piolibdeps/smartplug/AsyncTCP/src/AsyncTCP.cpp: In function 'void _send_async_event(lwip_tcp_event_packet_t*)':
.piolibdeps/smartplug/AsyncTCP/src/AsyncTCP.cpp:202:3: error: 'xTaskNotifyGive' was not declared in this scope
  202 |   xTaskNotifyGive(_async_service_task_handle);
      |   ^~~~~~~~~~~~~~~
.piolibdeps/smartplug/AsyncTCP/src/AsyncTCP.cpp: In function 'void _prepend_async_event(lwip_tcp_event_packet_t*)':
.piolibdeps/smartplug/AsyncTCP/src/AsyncTCP.cpp:210:3: error: 'xTaskNotifyGive' was not declared in this scope
  210 |   xTaskNotifyGive(_async_service_task_handle);
      |   ^~~~~~~~~~~~~~~
.piolibdeps/smartplug/AsyncTCP/src/AsyncTCP.cpp: In function 'void _async_service_task(void*)':
.piolibdeps/smartplug/AsyncTCP/src/AsyncTCP.cpp:322:5: error: 'ulTaskNotifyTake' was not declared in this scope
  322 |     ulTaskNotifyTake(pdTRUE, ASYNC_TCP_MAX_TASK_SLEEP);
      |     ^~~~~~~~~~~~~~~~
*** [.pioenvs/smartplug/lib052/AsyncTCP/AsyncTCP.cpp.o] Error 1

@vladT0 That does not work for me:

```
INFO ESPHome 2025.7.4
INFO Reading configuration /config/solis-modbus-wifi-stick.yaml...
INFO Generating C++ source...
INFO Compiling app...
Processing solis-modbus-wifi-stick (board: generic-rtl8710bx-4mb-980k; framework: arduino; platform: libretiny @ 1.6.0)
```

```
In file included from src/esphome/components/web_server_base/web_server_base.h:11,
                 from src/esphome/components/web_server_base/web_server_base.cpp:1:
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:463:75: error: 'emptyString' was not declared in this scope
  463 |     beginResponse(FS &fs, const String &path, const String &contentType = emptyString, bool download = false, AwsTemplateProcessor callback = nullptr) {
      |                                                                           ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:470:81: error: 'emptyString' was not declared in this scope
  470 |     beginResponse(File content, const String &path, const String &contentType = emptyString, bool download = false, AwsTemplateProcessor callback = nullptr) {
      |                                                                                 ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:684:77: error: 'emptyString' was not declared in this scope
  684 |   const String &getAttribute(const char *name, const String &defaultValue = emptyString) const;
      |                                                                             ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'const arduino::String& AsyncWebServerRequest::arg(int) const':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:575:20: error: 'emptyString' was not declared in this scope
  575 |     return i < 0 ? emptyString : arg((size_t)i);
      |                    ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'const arduino::String& AsyncWebServerRequest::argName(int) const':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:579:20: error: 'emptyString' was not declared in this scope
  579 |     return i < 0 ? emptyString : argName((size_t)i);
      |                    ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'const arduino::String& AsyncWebServerRequest::pathArg(int) const':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:591:20: error: 'emptyString' was not declared in this scope
  591 |     return i < 0 ? emptyString : pathArg((size_t)i);
      |                    ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'const arduino::String& AsyncWebServerRequest::header(int) const':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:606:20: error: 'emptyString' was not declared in this scope
  606 |     return i < 0 ? emptyString : header((size_t)i);
      |                    ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'const arduino::String& AsyncWebServerRequest::headerName(int) const':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:610:20: error: 'emptyString' was not declared in this scope
  610 |     return i < 0 ? emptyString : headerName((size_t)i);
      |                    ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'void AsyncWebServerRequest::setAttribute(const char*, bool)':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:668:39: error: 'emptyString' was not declared in this scope
  668 |     _attributes[name] = value ? "1" : emptyString;
      |                                       ^~~~~~~~~~~
In file included from src/esphome/components/web_server_base/web_server_base.h:11,
                 from src/esphome/components/web_server/web_server.h:5,
                 from src/esphome/components/web_server/list_entities.cpp:7:
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:463:75: error: 'emptyString' was not declared in this scope
  463 |     beginResponse(FS &fs, const String &path, const String &contentType = emptyString, bool download = false, AwsTemplateProcessor callback = nullptr) {
      |                                                                           ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:470:81: error: 'emptyString' was not declared in this scope
  470 |     beginResponse(File content, const String &path, const String &contentType = emptyString, bool download = false, AwsTemplateProcessor callback = nullptr) {
      |                                                                                 ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:684:77: error: 'emptyString' was not declared in this scope
  684 |   const String &getAttribute(const char *name, const String &defaultValue = emptyString) const;
      |                                                                             ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'const arduino::String& AsyncWebServerRequest::arg(int) const':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:575:20: error: 'emptyString' was not declared in this scope
  575 |     return i < 0 ? emptyString : arg((size_t)i);
      |                    ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'const arduino::String& AsyncWebServerRequest::argName(int) const':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:579:20: error: 'emptyString' was not declared in this scope
  579 |     return i < 0 ? emptyString : argName((size_t)i);
      |                    ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'const arduino::String& AsyncWebServerRequest::pathArg(int) const':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:591:20: error: 'emptyString' was not declared in this scope
  591 |     return i < 0 ? emptyString : pathArg((size_t)i);
      |                    ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'const arduino::String& AsyncWebServerRequest::header(int) const':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:606:20: error: 'emptyString' was not declared in this scope
  606 |     return i < 0 ? emptyString : header((size_t)i);
      |                    ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'const arduino::String& AsyncWebServerRequest::headerName(int) const':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:610:20: error: 'emptyString' was not declared in this scope
  610 |     return i < 0 ? emptyString : headerName((size_t)i);
      |                    ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'void AsyncWebServerRequest::setAttribute(const char*, bool)':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:668:39: error: 'emptyString' was not declared in this scope
  668 |     _attributes[name] = value ? "1" : emptyString;
      |                                       ^~~~~~~~~~~
In file included from .piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:1232,
                 from src/esphome/components/web_server_base/web_server_base.h:11,
                 from src/esphome/components/web_server_base/web_server_base.cpp:1:
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/WebResponseImpl.h: At global scope:
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/WebResponseImpl.h:25:83: error: 'emptyString' was not declared in this scope
   25 |   AsyncBasicResponse(int code, const String &contentType, const String &content = emptyString)
      |                                                                                   ^~~~~~~~~~~
In file included from src/esphome/components/web_server_base/web_server_base.h:11,
                 from src/esphome/components/web_server/web_server.h:5,
                 from src/esphome/components/web_server/web_server.cpp:1:
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:463:75: error: 'emptyString' was not declared in this scope
  463 |     beginResponse(FS &fs, const String &path, const String &contentType = emptyString, bool download = false, AwsTemplateProcessor callback = nullptr) {
      |                                                                           ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:470:81: error: 'emptyString' was not declared in this scope
  470 |     beginResponse(File content, const String &path, const String &contentType = emptyString, bool download = false, AwsTemplateProcessor callback = nullptr) {
      |                                                                                 ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:684:77: error: 'emptyString' was not declared in this scope
  684 |   const String &getAttribute(const char *name, const String &defaultValue = emptyString) const;
      |                                                                             ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'const arduino::String& AsyncWebServerRequest::arg(int) const':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:575:20: error: 'emptyString' was not declared in this scope
  575 |     return i < 0 ? emptyString : arg((size_t)i);
      |                    ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'const arduino::String& AsyncWebServerRequest::argName(int) const':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:579:20: error: 'emptyString' was not declared in this scope
  579 |     return i < 0 ? emptyString : argName((size_t)i);
      |                    ^~~~~~~~~~~
Compiling .pioenvs/solis-modbus-wifi-stick/src/esphome/core/scheduler.cpp.o
In file included from .piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:1232,
                 from src/esphome/components/web_server_base/web_server_base.h:11,
                 from src/esphome/components/web_server/web_server.h:5,
                 from src/esphome/components/web_server/list_entities.cpp:7:
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/WebResponseImpl.h: At global scope:
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/WebResponseImpl.h:25:83: error: 'emptyString' was not declared in this scope
   25 |   AsyncBasicResponse(int code, const String &contentType, const String &content = emptyString)
      |                                                                                   ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'const arduino::String& AsyncWebServerRequest::pathArg(int) const':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:591:20: error: 'emptyString' was not declared in this scope
  591 |     return i < 0 ? emptyString : pathArg((size_t)i);
      |                    ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'const arduino::String& AsyncWebServerRequest::header(int) const':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:606:20: error: 'emptyString' was not declared in this scope
  606 |     return i < 0 ? emptyString : header((size_t)i);
      |                    ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'const arduino::String& AsyncWebServerRequest::headerName(int) const':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:610:20: error: 'emptyString' was not declared in this scope
  610 |     return i < 0 ? emptyString : headerName((size_t)i);
      |                    ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'void AsyncWebServerRequest::setAttribute(const char*, bool)':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:668:39: error: 'emptyString' was not declared in this scope
  668 |     _attributes[name] = value ? "1" : emptyString;
      |                                       ^~~~~~~~~~~
Compiling .pioenvs/solis-modbus-wifi-stick/src/esphome/core/string_ref.cpp.o
In file included from .piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:1232,
                 from src/esphome/components/web_server_base/web_server_base.h:11,
                 from src/esphome/components/web_server/web_server.h:5,
                 from src/esphome/components/web_server/web_server.cpp:1:
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/WebResponseImpl.h: At global scope:
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/WebResponseImpl.h:25:83: error: 'emptyString' was not declared in this scope
   25 |   AsyncBasicResponse(int code, const String &contentType, const String &content = emptyString)
      |                                                                                   ^~~~~~~~~~~
In file included from src/esphome/components/web_server_base/web_server_base.h:11,
                 from src/esphome/components/web_server/web_server.h:5,
                 from src/esphome/components/web_server/web_server_v1.cpp:1:
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:463:75: error: 'emptyString' was not declared in this scope
  463 |     beginResponse(FS &fs, const String &path, const String &contentType = emptyString, bool download = false, AwsTemplateProcessor callback = nullptr) {
      |                                                                           ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:470:81: error: 'emptyString' was not declared in this scope
  470 |     beginResponse(File content, const String &path, const String &contentType = emptyString, bool download = false, AwsTemplateProcessor callback = nullptr) {
      |                                                                                 ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:684:77: error: 'emptyString' was not declared in this scope
  684 |   const String &getAttribute(const char *name, const String &defaultValue = emptyString) const;
      |                                                                             ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'const arduino::String& AsyncWebServerRequest::arg(int) const':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:575:20: error: 'emptyString' was not declared in this scope
  575 |     return i < 0 ? emptyString : arg((size_t)i);
      |                    ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'const arduino::String& AsyncWebServerRequest::argName(int) const':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:579:20: error: 'emptyString' was not declared in this scope
  579 |     return i < 0 ? emptyString : argName((size_t)i);
      |                    ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'const arduino::String& AsyncWebServerRequest::pathArg(int) const':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:591:20: error: 'emptyString' was not declared in this scope
  591 |     return i < 0 ? emptyString : pathArg((size_t)i);
      |                    ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'const arduino::String& AsyncWebServerRequest::header(int) const':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:606:20: error: 'emptyString' was not declared in this scope
  606 |     return i < 0 ? emptyString : header((size_t)i);
      |                    ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'const arduino::String& AsyncWebServerRequest::headerName(int) const':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:610:20: error: 'emptyString' was not declared in this scope
  610 |     return i < 0 ? emptyString : headerName((size_t)i);
      |                    ^~~~~~~~~~~
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h: In member function 'void AsyncWebServerRequest::setAttribute(const char*, bool)':
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:668:39: error: 'emptyString' was not declared in this scope
  668 |     _attributes[name] = value ? "1" : emptyString;
      |                                       ^~~~~~~~~~~
In file included from .piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/ESPAsyncWebServer.h:1232,
                 from src/esphome/components/web_server_base/web_server_base.h:11,
                 from src/esphome/components/web_server/web_server.h:5,
                 from src/esphome/components/web_server/web_server_v1.cpp:1:
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/WebResponseImpl.h: At global scope:
.piolibdeps/solis-modbus-wifi-stick/ESPAsyncWebServer/src/WebResponseImpl.h:25:83: error: 'emptyString' was not declared in this scope
   25 |   AsyncBasicResponse(int code, const String &contentType, const String &content = emptyString)
      |                                                                                   ^~~~~~~~~~~
*** [.pioenvs/solis-modbus-wifi-stick/src/esphome/components/web_server/list_entities.cpp.o] Error 1
*** [.pioenvs/solis-modbus-wifi-stick/src/esphome/components/web_server_base/web_server_base.cpp.o] Error 1
*** [.pioenvs/solis-modbus-wifi-stick/src/esphome/components/web_server/web_server_v1.cpp.o] Error 1
*** [.pioenvs/solis-modbus-wifi-stick/src/esphome/components/web_server/web_server.cpp.o] Error 1
```

I commented, but the same error

`Compiling` .pioenvs/smartplug/liba5f/ESPAsyncWebServer/AsyncWebHeader.cpp.o
.piolibdeps/smartplug/AsyncTCP/src/AsyncTCP.cpp: In function 'void _send_async_event(lwip_tcp_event_packet_t*)':
.piolibdeps/smartplug/AsyncTCP/src/AsyncTCP.cpp:202:3: error: 'xTaskNotifyGive' was not declared in this scope
  202 |   xTaskNotifyGive(_async_service_task_handle);
      |   ^~~~~~~~~~~~~~~
.piolibdeps/smartplug/AsyncTCP/src/AsyncTCP.cpp: In function 'void _prepend_async_event(lwip_tcp_event_packet_t*)':
.piolibdeps/smartplug/AsyncTCP/src/AsyncTCP.cpp:210:3: error: 'xTaskNotifyGive' was not declared in this scope
  210 |   xTaskNotifyGive(_async_service_task_handle);
      |   ^~~~~~~~~~~~~~~
.piolibdeps/smartplug/AsyncTCP/src/AsyncTCP.cpp: In function 'void _async_service_task(void*)':
.piolibdeps/smartplug/AsyncTCP/src/AsyncTCP.cpp:322:5: error: 'ulTaskNotifyTake' was not declared in this scope
  322 |     ulTaskNotifyTake(pdTRUE, ASYNC_TCP_MAX_TASK_SLEEP);
      |     ^~~~~~~~~~~~~~~~
*** [.pioenvs/smartplug/lib052/AsyncTCP/AsyncTCP.cpp.o] Error 1
========================= [FAILED] Took 374.37 seconds =========================

--- a/esphome/components/rtl87xx/__init__.py
+++ b/esphome/components/rtl87xx/__init__.py
@@ -6,6 +6,7 @@
 # in schema.py file in this directory.
 
 from esphome import pins
+import esphome.codegen as cg
 from esphome.components import libretiny
 from esphome.components.libretiny.const import (
     COMPONENT_RTL87XX,
@@ -45,6 +46,9 @@ CONFIG_SCHEMA.prepend_extra(_set_core_data)
 
 
 async def to_code(config):
+    # Use FreeRTOS 8.2.3+ for xTaskNotifyGive/ulTaskNotifyTake required by AsyncTCP 3.4.3+
+    # https://github.com/esphome/esphome/issues/10220
+    cg.add_platformio_option("custom_versions.freertos", "8.2.3")
     return await libretiny.component_to_code(config)
 
 

